Userland/Libraries/LibJS/Runtime/DisposableStack.h

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright (c) 2022, David Tuin <davidot@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <LibJS/Runtime/AbstractOperations.h>
+#include <LibJS/Runtime/Object.h>
+
+namespace JS {
+
+class DisposableStack final : public Object {
+    JS_OBJECT(DisposableStack, Object);
+
+public:
+    virtual ~DisposableStack() override = default;
+
+    enum class DisposableState {
+        Pending,
+        Disposed
+    };
+
+    [[nodiscard]] DisposableState disposable_state() const { return m_state; }
+    Vector<DisposableResource> const& disposable_resource_stack() const { return m_disposable_resource_stack; }
+    Vector<DisposableResource>& disposable_resource_stack() { return m_disposable_resource_stack; }
```

Comment:
I think all new getters should be nodiscard (there might be some in earlier commits that I missed)